### PR TITLE
Template Pipeline: Generate Temporary Variables

### PR DESCRIPTION
**Pull request by dylhunn**
The expression `a()?.b` should expand into `(tmp = a()) === null ? null : tmp.b`, in order to avoid calling the function `a()` twice.

This commit modifies the null-safe-expansion algorithm to emit temporary assignments, and provides the reification code to actually generate the declarations, assignments, and reads.

Note also that, with our bottom-up algorithm, there are some tricky cases when a function call exists inside an indexed access, such as `f1()?.[f2()?.a]?.b`. We add some special logic to avoid generating a double-assignment to the temporary storing the result of `f2()`.

Finally, there are opportunities to reuse the same temporary in expressions like `a?.[f()]?.[f()]`. We save this for the next commit.

### CODE REVIEW

**Comment by eneajaho** [comment]
I have a question: 

What happens if the function has side effects? Or doesn't always return the same value. 

Nit: This is an absurd case, but it's a case 😅

**Comment by dylhunn** [comment]
> I have a question:
> 
> What happens if the function has side effects? Or doesn't always return the same value.
> 
> Nit: This is an absurd case, but it's a case 😅

@eneajaho Due to this transformation, we call the function exactly once, which is what you might expect if you write `a()?.b`. So the behavior should not change. (Without this transformation, we'd cause the side effects to happen twice.)

File: packages/compiler/src/template/pipeline/src/phases/expand_safe_reads.ts [hunk 1]
```diff
@@ -10,30 +10,114 @@ import * as o from '../../../../output/output_ast';
...
+    return true;
+  } else if (e instanceof o.LiteralArrayExpr) {
+    return true;
+  } else if (e instanceof o.LiteralMapExpr) {
+    return true;
+  } else if (e instanceof ir.PipeBindingExpr) {
+    // TODO: Is this the right way to detect a pipe in the expression?
+    return true;
+  }
+  return false;
```
Comment (alxhub) [inline packages/compiler/src/template/pipeline/src/phases/expand_safe_reads.ts:62]: We should not have default behaviors for expression types - let's list everything out.

File: packages/compiler/src/template/pipeline/src/phases/reify.ts [hunk 2]
```diff
@@ -172,6 +172,16 @@ function reifyIrExpression(expr: o.Expression): o.Expression {
...
+    case ir.ExpressionKind.ReadTemporaryExpr:
+      if (expr.name === null) {
+        throw new Error(`Read of unnamed temporary ${expr.xref}`);
+      }
+      return o.variable(expr.name);
+    case ir.ExpressionKind.AssignTemporaryExpr:
+      if (expr.name === null) {
+        throw new Error(`Assign of unnamed temporary ${expr.xref}`);
+      }
+      return new o.WriteVarExpr(expr.name, expr.expr);
```
Comment (alxhub) [inline packages/compiler/src/template/pipeline/src/phases/reify.ts:184]:
```suggestion
      o.variable(expr.name).set(expr.expr);
```

File: packages/compiler/src/template/pipeline/src/phases/expand_safe_reads.ts [hunk 3]
```diff
@@ -10,30 +10,114 @@ import * as o from '../../../../output/output_ast';
 import * as ir from '../../ir';
 import {ComponentCompilation} from '../compilation';
 
+type SafeTransformContext = {
```
Comment (alxhub) [inline packages/compiler/src/template/pipeline/src/phases/expand_safe_reads.ts:13]:
```suggestion
interface SafeTransformContext {
```

File: packages/compiler/src/template/pipeline/src/phases/expand_safe_reads.ts [hunk 4]
```diff
@@ -10,30 +10,114 @@ import * as o from '../../../../output/output_ast';
...
+      // temporary refers to, possibly through nested temporaries, has a function call. We copy that
+      // behavior here.
+      return ctx.compatibility ? new ir.AssignTemporaryExpr(read, read.xref) : read;
+    }
+    return e;
+  }, ir.VisitorContextFlag.None);
+  return e;
+}
+
+function maybeWithTemporary(e: o.Expression, ctx: SafeTransformContext) {
```
Comment (alxhub) [inline packages/compiler/src/template/pipeline/src/phases/expand_safe_reads.ts:97]: Declare return type

File: packages/compiler/src/template/pipeline/src/phases/expand_safe_reads.ts [hunk 5]
```diff
@@ -10,30 +10,114 @@ import * as o from '../../../../output/output_ast';
...
+  let result: [o.Expression, o.Expression];
+  if (needsTemporaryInSafeAccess(e)) {
+    const xref = ctx.cpl.allocateXrefId();
+    return [new ir.AssignTemporaryExpr(e, xref), new ir.ReadTemporaryExpr(xref)];
+  }
+  result = [e, e.clone()];
+  eliminateTemporaryAssignments(result[1], temporariesIn(result[0]), ctx);
```
Comment (alxhub) [inline packages/compiler/src/template/pipeline/src/phases/expand_safe_reads.ts:104]: At least a comment here about _why_ this happens

**Review by alxhub — APPROVED** [review]

**Comment by dylhunn** [comment]
caretaker: windows target is a flake. g3 is not affected.

**Comment by alfaproject** [comment]
OOC are there plans to make these the same as JS semantics?

**Comment by dylhunn** [comment]
@AndrewKushnir PR is rebased.

**Pull request closed without merging**